### PR TITLE
Fix omissions in GridGenerator::flatten_triangulation().

### LINKED ISSUE #13126: GridGenerator::flatten_triangulation() does not copy manifold ids of internal objects
As mentioned in the comments of #13117, `GridGenerator::flatten_manifold()` does not copy the manifold ids of internal objects.

### LINKED ISSUE #13127: GridGenerator::flatten_triangulation() does not copy data on boundary edges in 3d
Another issue I found while doing #13117 is that in 3d, the function only copies information about boundary *faces*, but not *edges*. That's a bug.

### PR DESCRIPTION
**Pull request by bangerth**
This is a final (?) follow-up to #13117, #13128, and related patches. It fixes the two bugs identified in #13126 and #13127, namely the omission of edges in 3d and the omission of copying manifold ids of internal faces and edges.

Fixes #13126 . Fixes #13127 .

/rebuild

### CODE REVIEW

**Comment by bangerth** [comment]
This patch has #13142 as part of it. I will rebase once #13142 is merged, but the patch can be reviewed as is already.

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
...
         case 2:
           {
-            for (const auto &face : in_tria.active_face_iterators() |
-                                      IteratorFilters::AtBoundary())
+            std::vector<bool> user_flags_line;
+            in_tria.save_user_flags_line(user_flags_line);
+            const_cast<Triangulation<dim, spacedim1> &>(in_tria)
+              .clear_user_flags_line();
+
+            // Loop over the boundary faces of the triangulation and create
```
Comment (masterleinad) [inline source/grid/grid_generator.cc:7548]:
```suggestion
            // Loop over all the faces of the triangulation and create
```

File: source/grid/grid_generator.cc [hunk 2]
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
...
+            in_tria.save_user_flags_line(user_flags_line);
+            const_cast<Triangulation<dim, spacedim1> &>(in_tria)
+              .clear_user_flags_line();
+
+            std::vector<bool> user_flags_quad;
+            in_tria.save_user_flags_quad(user_flags_quad);
+            const_cast<Triangulation<dim, spacedim1> &>(in_tria)
+              .clear_user_flags_quad();
+
+            // Loop over the boundary faces of the triangulation and create
```
Comment (masterleinad) [inline source/grid/grid_generator.cc:7614]:
```suggestion
            // Loop over all the faces of the triangulation and create
```

File: source/grid/grid_generator.cc [hunk 3]
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
...
+                        boundary_face.manifold_id = face->manifold_id();
+
+                        subcelldata.boundary_quads.emplace_back(
+                          std::move(boundary_face));
+
+                        face->set_user_flag();
+                      }
+
+                    // Then also loop over the edges of this face. Because every
+                    // edge must also be a part of a boundary face, we can
```
Comment (masterleinad) [inline source/grid/grid_generator.cc:7691]:
```suggestion
                    // boundary edge must also be a part of a boundary face, we can
```

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Comment by bangerth** [comment]
#13142 is now merged, but the testers reveal that the MS VC compiler wants certain functions to be instantiated. That required me to work around an overload issue, so while I have dropped the commits that were part of #13142 and rebased, I needed to add two more commits to make things work. Let's see whether the MS VC compiler is happy.

**Comment by bangerth** [comment]
It works now.

**Comment by bangerth** [comment]
Ping?

**Review by Rombur — APPROVED** [review]

**Pull request merged**